### PR TITLE
BIP-X: Remove Withdraw Timer

### PR DESCRIPTION
**Pull request by Brean0**
# BIP-X: Remove Withdraw Timer

- [Proposer](#proposer)
- [Summary](#summary)
- [Problem](#problem)
- [Proposed Solution](#proposed-solution)

## Proposer

Beanstalk Farms

## Summary

- Adjust the issuance of earned Beans such that it is linearly distributed throughout the season, rather than at the start of the season.
- Remove the Withdrawal Freeze, allowing farmers to deposit and withdraw beans in the same season.
- Deprecate the various withdraw functions. 

## Problem

Numerous inefficiencies remain within Beanstalk: 

1. Beanstalk cannot currently support decrementing the Withdrawal timer to 0 because anyone could then receive “risk-free” Earned Beans because Earned Beans are distributed instantaneously when the sunrise() function is called. 
2. Having a withdraw lockup creates an unnecessarily complex UX as Farmers are forced to perform 2 on-chain actions at different times. This limits the potential for using Deposited Beans throughout DeFi via Pipeline as Deposited Beans cannot atomically be swapped for Beans.

## Proposed Solution

- Emit newly minted Earned Beans to Stalkholders proportional to the percent of the Season that has elapsed instead of instantly on the sunrise call. 
     - https://github.com/BeanstalkFarms/Beanstalk/issues/149
- Remove the Withdrawal Freeze entirely, allowing Farmers to Withdraw directly to Circulating/Farm balances in a single transaction.
     - https://github.com/BeanstalkFarms/Beanstalk/issues/150 

 
## Diagrams/Visualizations
image
image

### CODE REVIEW

File: protocol/contracts/beanstalk/farm/FarmFacet.sol [hunk 1]
```diff
@@ -5,9 +5,9 @@
 pragma solidity ^0.7.6;
 pragma experimental ABIEncoderV2;
 
+import {LibDiamond} from "~/libraries/LibDiamond.sol";
+import {LibEth} from "~/libraries/Token/LibEth.sol";
```
Comment (funderberker) [inline protocol/contracts/beanstalk/farm/FarmFacet.sol:9]: why did we change the import style and is it supposed to be changed to the tilde style everywhere?

File: protocol/contracts/beanstalk/init/replant/Replant1.sol [hunk 2]
```diff
@@ -82,7 +82,7 @@ contract Replant1 {
 
     function init() external {
         // 1. Remove Deposits
-        LibTokenSilo.removeDeposit(
+        LibTokenSilo.removeDepositFromAccount(
```
Comment (funderberker) [inline protocol/contracts/beanstalk/init/replant/Replant1.sol:85]: Do we need to keep these one time use legacy files up to date with changes? Seems best to keep them in the state that they were used in.

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 3]
```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
...
+        // We round up as it prevents an account having roots but no stalk.
+        
+        // if the user withdraws in the same block as sunrise, they forfeit their earned beans for that season
+        // this is distrubuted to the other users.
+        // should this be the same as the vesting period?
+        if(block.number - s.season.sunriseBlock <= 25){
+            roots = s.s.roots.mulDiv(
+            stalk,
+            s.s.stalk-s.newEarnedStalk,
+            LibPRBMath.Rounding.Up);
```
Comment (funderberker) [inline protocol/contracts/libraries/Silo/LibSilo.sol:224]: @Brean0 previous logic was using safe math that rounds towards zero. why the change here and what are the implications?

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 4]
```diff
@@ -60,20 +69,50 @@ contract Account {
...
+         * 
+         * @dev FIXME(doc)
+         */
         uint32 votedUntil; // DEPRECATED – Replant removed on-chain governance including the ability to vote on BIPs.
         uint32 lastUpdate; // The Season in which the Farmer last updated their Silo.
         uint32 lastSop; // The last Season that a SOP occured at the time the Farmer last updated their Silo.
         uint32 lastRain; // The last Season that it started Raining at the time the Farmer last updated their Silo.
-        uint32 lastSIs; // DEPRECATED – In Silo V1.2, the Silo reward mechanism was updated to no longer need to store the number of the Supply Increases at the time the Farmer last updated their Silo.
-        uint32 proposedUntil; // DEPRECATED – Replant removed on-chain governance including the ability to propose BIPs.
+        uint128 deltaRoots; // the number of roots to add, in the case where a farmer has mowed in the morning
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/AppStorage.sol:110]: in the EVM, data are stored in 32 byte slots. This allows for packing multiple variables into 1 slot, and adding more variables without overwriting the storage slot, as long the new variables do not exceed the 32 byte length. 

In this case, previously 6 uint32s were stored in this slot. This means that 24 bytes were used, and a potential 8 bytes can be added to this slot without affecting the leading slots after. 

by removing the last 2 uint32 variables in the slot, we can now pack a uint128 (16 bytes) into the slot without affecting the leading slots. The slot now looks like: 
uint32  (4 bytes)
uint32  (4 bytes)
uint32  (4 bytes)
uint32  (4 bytes) 
uint128 (16 bytes)
(4+4+4+4+12 = 32) 
additionally, because delta roots is only used in the vesting period, and is always updated before a vesting period plant, the previous values that was contained in the variables will be overwritten and not used.

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 5]
```diff
@@ -235,12 +294,12 @@ contract Storage {
...
 
 struct AppStorage {
-    uint8 index; // DEPRECATED - Was the index of the Bean token in the Bean:Eth Uniswap v2 pool, which has been depreciated.
+    uint8 index; // DEPRECATED - Was the index of the Bean token in the BEAN:ETH Uniswap v2 pool, which has been deprecated.
     int8[32] cases; // The 24 Weather cases (array has 32 items, but caseId = 3 (mod 4) are not cases).
     bool paused; // True if Beanstalk is Paused.
     uint128 pausedAt; // The timestamp at which Beanstalk was last paused. 
     Storage.Season season; // The Season storage struct found above.
-    Storage.Contracts c; // DEPRECATED - Previously stored the Contracts State struct. Removed when contract addresses were moved to constants in C.sol.
+    uint256[4] deprecated; // DEPRECATED - Previously stored the Contracts State struct. Removed when contract addresses were moved to constants in C.sol.
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/AppStorage.sol:302]: since data is stored in 32 byte chunks, `address[4]`  still used 4 slots of storage, even though the array technically only used (4 * 20) 80 bytes. thus 4 slots are deprecated here (uint256[4] and bytes32[4] are the same in this case)

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 6]
```diff
@@ -284,6 +358,7 @@ struct AppStorage {
     uint128 fFirst; // The lowest active Fertilizer Id (start of linked list that is stored by nextFid). 
     uint128 fLast; // The highest active Fertilizer Id (end of linked list that is stored by nextFid). 
     uint128 bpf; // The cumulative Beans Per Fertilizer (bfp) minted over all Season.
+    uint128 newEarnedRoots; // the number of roots to add to the global roots, in the case the user plants in the morning. // placed here to save a storage slot.s
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/AppStorage.sol:361]: bpf slot previously: 
uint128 (16 bytes) 
16 free bytes -> newEarnedRoots

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 7]
```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
...
+
         uint256 roots;
-        if (s.s.roots == 0) roots = stalk.mul(C.getRootsBase());
-        else roots = s.s.roots.mul(stalk).div(s.s.stalk);
+        if (s.s.roots == 0) {
+            roots = uint256(stalk.mul(C.getRootsBase()));
+        } else  {
+            roots = s.s.roots.mul(stalk).div(s.s.stalk);
+            if (block.number - s.season.sunriseBlock <= 25) {
+                uint256 rootsWithoutEarned = s.s.roots.add(s.newEarnedRoots).mul(stalk).div(s.s.stalk - (s.newEarnedStalk));
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibSilo.sol:158]: `newEarnedRoots` are not accounted for in s.s.roots, as roots are not incremented when new beans are minted, but rather during the mow function.

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 8]
```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
...
-        uint256 roots = s.s.roots.mul(stalk).div(s.s.stalk);
+       
+        uint256 roots;
+        // Calculate the amount of Roots for the given amount of Stalk.
+        // We round up as it prevents an account having roots but no stalk.
+        
+        // if the user withdraws in the same block as sunrise, they forfeit their earned beans for that season
+        // this is distrubuted to the other users.
+        // should this be the same as the vesting period?
+        if(block.number - s.season.sunriseBlock <= 25){
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibSilo.sol:220]: agree

File: protocol/contracts/libraries/Silo/LibSilo.sol (hunk 3 above)
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibSilo.sol:224]: previously the roots here was truncated - with the introduction of more complex math, this adds an additional fail safe to prevent rounding errors. This follows defi standards (we round down when issuing roots, round up when removing roots) to prevent the protocol from being undercollateralized

Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibSilo.sol:224]: Additionally, line 95 adds an additional precaution to prevent underflow.

File: protocol/contracts/libraries/Silo/LibTokenSilo.sol [hunk 9]
```diff
@@ -25,172 +34,266 @@ library LibTokenSilo {
...
+     *
+     * FIXME(discuss): If we think of Deposits like 1155s, we might call the
+     * combination of "incrementTotalDeposited" and "addDepositToAccount" as 
+     * "minting a deposit".
+     */
     function depositWithBDV(
         address account,
         address token,
-        uint32 _s,
+        uint32 season,
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibTokenSilo.sol:92]: season >>>> _s

File: protocol/contracts/beanstalk/AppStorage.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (funderberker) [reply protocol/contracts/beanstalk/AppStorage.sol:110]: got it. ty for the explanation

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloExit.sol [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+        view
+        returns (uint256 beans) {
+        // There will be no Roots before the first Deposit is made.
+        if (s.s.roots == 0) return 0;
+
+        uint256 stalk;
+        if(block.number - s.season.sunriseBlock <= EARNED_BEAN_VESTING_BLOCKS){
+            stalk = s.s.stalk.sub(s.newEarnedStalk).mulDiv(
+                s.a[account].roots.add(s.a[account].deltaRoots), // add the delta roots of the user
+                s.s.roots.add(s.newEarnedRoots), // add delta of global roots
```
Comment (funderberker) [inline protocol/contracts/beanstalk/silo/SiloFacet/SiloExit.sol:192]: s.newEarnedBeans is set when calling mow, which may not be called before checking of balance here. should we trigger an update before this addition?

File: protocol/contracts/libraries/Silo/LibSilo.sol (hunk 3 above)
Comment (funderberker) [reply protocol/contracts/libraries/Silo/LibSilo.sol:224]: what is the precaution? line 95 is `mintStalk(account, stalk); // also mints Roots`

Comment (funderberker) [reply protocol/contracts/libraries/Silo/LibSilo.sol:224]: let's talk about rounding on the next call. i think i am missing a lot of context

File: protocol/contracts/libraries/Silo/LibSilo.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (funderberker) [reply protocol/contracts/libraries/Silo/LibSilo.sol:158]: ok, but why are they being added here? this logic is (roots with earned) * (stalk ratio without earned)
but seems like it should be (roots without earned) * (stalk ratio without earned)

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 11]
```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
...
-        if (s.s.roots == 0) roots = stalk.mul(C.getRootsBase());
-        else roots = s.s.roots.mul(stalk).div(s.s.stalk);
+        if (s.s.roots == 0) {
+            roots = uint256(stalk.mul(C.getRootsBase()));
+        } else  {
+            roots = s.s.roots.mul(stalk).div(s.s.stalk);
+            if (block.number - s.season.sunriseBlock <= 25) {
+                uint256 rootsWithoutEarned = s.s.roots.add(s.newEarnedRoots).mul(stalk).div(s.s.stalk - (s.newEarnedStalk));
+                uint256 deltaRoots = rootsWithoutEarned - roots;
+                s.newEarnedRoots = s.newEarnedRoots.add(uint128(deltaRoots));
```
Comment (funderberker) [inline protocol/contracts/libraries/Silo/LibSilo.sol:160]: this `deltaRoots` is the delta for the account stalk amount entered as parameters here. is it intentional that the silo state `newEarnedRoots` is holding a value for an account?

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 12]
```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
...
+                s.a[account].deltaRoots = uint128(deltaRoots);
+            } 
+        }
 
+        // increment user and total stalk
         s.s.stalk = s.s.stalk.add(stalk);
         s.a[account].s.stalk = s.a[account].s.stalk.add(stalk);
 
+        // increment user and total roots
         s.s.roots = s.s.roots.add(roots);
```
Comment (funderberker) [inline protocol/contracts/libraries/Silo/LibSilo.sol:170]: this roots incrementation includes "new earned roots" is that intentional?

File: protocol/test/Silo.test.js [hunk 13]
```diff
@@ -133,5 +144,302 @@ describe('Silo', function () {
       await this.silo.connect(user2).plant()
       expect(await this.silo.totalEarnedBeans()).to.eq('0');
     });
-  })
+  });
+  
+
+  describe("Time Weighted Earned Bean Emission", async function () {
```
Comment (funderberker) [inline protocol/test/Silo.test.js:150]: reminder: update name

File: protocol/contracts/beanstalk/init/replant/Replant1.sol (hunk 2 above)
Comment (funderberker) [reply protocol/contracts/beanstalk/init/replant/Replant1.sol:85]: @silochad, do you have an opinion?

**Comment by silochad** [comment]
Closing as this has been manually merged into `grown-stalk-per-bdv`.
